### PR TITLE
Fixed #1682 -- alert user when using file field without proper encoding

### LINKED ISSUE #1682: Log when form is using file fields without proper enctype attribute on the form.
One of the issues that can trip up new developers is the fact that forms require `enctype="multipart/form-data"` when there's a file being uploaded. If a developer doesn't know this, it can be hard to track down. I think it'd be reasonable to have the toolbar inspect the loaded HTML content and inspect it for invalid configurations. The first of which being a `<input type="file" ..> />` in a form that isn't defined as `<form enctype="multipart/form-data">`.

### PR DESCRIPTION
**Pull request by bkdekoning**
# Description

Added functionality where the toolbar inspects the loaded HTML content for a form that includes a file input but does not have the encoding type set to `multipart/form-data`, and warns the user if so. This is implemented in a separate alerts panel, as discussed. The alerts panel is placed at the top of the toolbar, to ensure that it receives appropriate attention when issues arise. My one concern with this implementation is that panel may cause people to expect that the toolbar checks for a wide range of issues, whereas it currently just checks for a single one.

What I did specifically is subclass Python's HTML parser and added a function called `check_invalid_file_form_configuration` in `panels/templates/panel.py`. The function checks all forms for a file input type and, if it exists, checks if the form has encoding type multipart/form-data. If it does not, the error message is shown in the alerts panel by passing it to `record_stats`. It also amends the `nav_subtitle` property of the panel to notify the user that there is an issue. Added four tests and checked that the test suite passes for all compatible versions of Django and Python through Tox.

Fixes #1682

# Checklist:

- [x] I have added the relevant tests for this change.
- [x] I have added an item to the Pending section of ``docs/changes.rst``.

### CODE REVIEW

**Comment by tim-schilling** [comment]
By the way, you don't have to recreate the PRs. I think for 95% of cases, we're happy with re-using the same PR for a change.

File: debug_toolbar/panels/alerts.py
Comment (tim-schilling) [inline debug_toolbar/panels/alerts.py:78]: We should also check that the [`form` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#form) is not set or set to an empty string. We could potentially handle this, but it's pretty un-django-ific in my opinion. Basically if someone is using that attribute, then let's disable this check for that input.

File: debug_toolbar/panels/alerts.py
Comment (tim-schilling) [inline debug_toolbar/panels/alerts.py:54]: I wonder if it'd be better to set this in [`enable_instrumentation`](https://django-debug-toolbar.readthedocs.io/en/latest/panels.html#debug_toolbar.panels.Panel.enable_instrumentation) instead

File: debug_toolbar/panels/alerts.py
Comment (tim-schilling) [inline debug_toolbar/panels/alerts.py:65]: Since the panel is called `AlertsPanel`, let's use `alerts` rather than `issues`.

File: debug_toolbar/panels/alerts.py
Comment (tim-schilling) [inline debug_toolbar/panels/alerts.py:67]: Let's include your great description from your commit in the doc string here:

> Inspects HTML content for a form that includes a file input but does not have the encoding type set to multipart/form-data, and warns the user if so.

File: docs/panels.rst
Comment (tim-schilling) [inline docs/panels.rst:19]:
```suggestion
This panel shows alerts for a set of pre-defined cases:

- Alerts when the response has a form without the
  ``enctype="multipart/form-data"`` attribute and the form contains
  a file input
```

I think it's prudent to start listing them off from the start.

File: debug_toolbar/panels/alerts.py
Comment (tim-schilling) [inline debug_toolbar/panels/alerts.py:58]: Please use `get_stats` to find the alerts rather than relying on the instance's member. It'll remove a step for us with the async project.

**Review by tim-schilling — CHANGES_REQUESTED** [review]
I've got a few opinionated takes on the approach. Let me know if I'm overreaching with any of them. I really like what you have here. I think changing the `nav_subtitle` is a great first version for notifying the user of a problem! Nice solution there.

Well done with the tests, documentation and arrangement of the code too!

File: debug_toolbar/panels/alerts.py
Comment (bkdekoning) [reply debug_toolbar/panels/alerts.py:54]: I trust your call on this, but have to admit that I do not fully understand the `enable_instrumentation` method. What would the advantage be of setting it there? Looking at the other panels, I commonly see lists being initialized in `__init__`.

File: debug_toolbar/panels/alerts.py
Comment (bkdekoning) [reply debug_toolbar/panels/alerts.py:78]: I'm happy to try and handle this. We can check for the existence of `tag == "input" and attrs.get("form")` and then check the encoding of the referenced forms by their IDs. Would that be an appropriate solution in your view?

**Comment by bkdekoning** [comment]
While working on the above, I realized that I am implicitly assuming that the form is not nested. While forms [should not be nested](https://html.spec.whatwg.org/multipage/forms.html#the-form-element), I tested such a construction with Chrome, Firefox, and Edge, and they handle it by simply ignoring the "child" form. 

As such, we might end up in a situation where we are not alerting the developer to an issue that may exist for some users -- when the encoding is properly set in the "child" form but not the "parent" form. Is this something we want to issue an additional alert for or would that be overly cautious?

File: debug_toolbar/panels/alerts.py
Comment (bkdekoning) [reply debug_toolbar/panels/alerts.py:102]: I have implemented this.

**Comment by tim-schilling** [comment]
As a heads up, I probably won't get to this towards the end of June. Things are ramping up elsewhere and the toolbar is taking a back seat. Thank you for making the changes so far!

**Comment by bkdekoning** [comment]
Alright, thanks for the heads-up. For future reference, I think the remaining items to settle on are: 
- Whether to use `enable_instrumentation` or `__init__` to initialize alerts
- Whether we should handle direct form references in file inputs
- Whether we should handle nested forms

Good luck with your other activities!

**Comment by tim-schilling** [comment]
> Whether we should handle nested forms

Nested forms are against the HTML spec, so I think what we have is fine. Most people aren't relying on this to be perfect, so let's get something "good" out.

> Whether we should handle direct form references in file inputs

While I'd like to see us do that, I don't think we have to worry too much about it. It would take a fair amount of work to get a Django form renderer to handle that. Let's lean into what's the common flow for Django devs.

**Comment by tim-schilling** [comment]
> Whether to use enable_instrumentation or __init__ to initialize alerts

Looking at this again, it's not a big deal. Let's go with what you have.

**Review by tim-schilling — APPROVED** [review]
Code looks great. I'm going to pull it down and check the UI to see what it looks like. If that's good, I'll merge it.

**Comment by bkdekoning** [comment]
> > Whether we should handle direct form references in file inputs
> 
> While I'd like to see us do that, I don't think we have to worry too much about it. It would take a fair amount of work to get a Django form renderer to handle that. Let's lean into what's the common flow for Django devs.

Just to be sure: I did implement this. When the parser sees a file input field that directly references a form, it adds it to the `referenced_file_inputs` list. The `check_invalid_file_form_configuration` function then fetches the referenced form's encoding type and checks if it is set to `multipart/form-data`.

**Comment by tim-schilling** [comment]
Oh, perfect!

Take a look at the latest commits and let me know if you have any issues with them. The biggest changes were moving the panel and rewording the messages a bit.

**Comment by bkdekoning** [comment]
All looks good to me!

**Comment by tim-schilling** [comment]
@matthiask I'd like to get your approval on the new panel before we merge this.

**Comment by matthiask** [comment]
Yes sure, let's do it!

**Comment by bkdekoning** [comment]
Thank you, @tim-schilling and @matthiask! Very excited to have made my first contribution. I really appreciated the thorough review and guidance.

**Comment by matthiask** [comment]
I haven't dont much here, the thank you belongs to you and Tim!

**Pull request merged**
